### PR TITLE
Reorder list based on map view

### PR DESCRIPTION
**Pull request by pezholio**
Fixes mysociety/fixmystreetforcouncils#152

This loops through all the markers on a map, and, if they're visible, appends the associated list item to another `<ul>` element above the main `.item-list` element, effectively showing the visible elements first. I've added an extra `id` attribute to each item now to, as this is considerably faster than searching by `data` element.

### CODE REVIEW

**Comment by pezholio** [comment]
Quick example of how it should look:

avllisfo3o

**Comment by pezholio** [comment]
Thoughts welcome on the UI here. I could append the visible items to the existing list container, but I think the border acts as a useful visual cue as to the state of the other reports.

**Comment by pezholio** [comment]
I've added this functionality to the planned reports pages, but this messes with the reordering functionality. I'm also wondering if this is a useful thing to have at all on the shortlist page, given that there is likely to be fewer than c. 20 results on these pages

File: web/js/map-OpenLayers.js [hunk 1]
```diff
@@ -115,6 +115,29 @@ var fixmystreet = fixmystreet || {};
...
+      reorder_visible: function() {
+          var visible = ''
+          var $itemList = $('.item-list');
+          var $visibleReports = $('.visible-reports');
+          for (var i = 0; i < fixmystreet.markers.features.length; i++) {
+              var feature = fixmystreet.markers.features[i];
+              var $listItem = $itemList.find('#report-'+ feature.data.id);
+              if (feature.onScreen()) {
+                  visible += $listItem[0].outerHTML; // This is faster than using .append()
+                  $listItem.addClass('hidden');
```
Comment (dracos) [inline web/js/map-OpenLayers.js]: On an `/around` page, this hides all the reports, I'm afraid. (As there's no visible-reports div, so they all get hidden from the main div.)
I'm not sure this is appropriate behaviour for the around page, which is already doing "what's on the map" (plus a few outside to make sure something is shown), should it be restricted to the reports listing page only?

File: web/js/map-OpenLayers.js [hunk 2]
```diff
@@ -115,6 +115,29 @@ var fixmystreet = fixmystreet || {};
...
 
+      reorder_visible: function() {
+          var visible = ''
+          var $itemList = $('.item-list');
+          var $visibleReports = $('.visible-reports');
+          for (var i = 0; i < fixmystreet.markers.features.length; i++) {
+              var feature = fixmystreet.markers.features[i];
+              var $listItem = $itemList.find('#report-'+ feature.data.id);
+              if (feature.onScreen()) {
+                  visible += $listItem[0].outerHTML; // This is faster than using .append()
```
Comment (dracos) [inline web/js/map-OpenLayers.js]: This means each visible report is now in the DOM twice with the same ID. As long as nothing is relying on this (it's new here, so presumably not), I guess that's okay, but the worry is if something comes along and assumes it's unique in future.

File: web/js/map-OpenLayers.js [hunk 3]
```diff
@@ -115,6 +115,29 @@ var fixmystreet = fixmystreet || {};
         fixmystreet.markers.redraw();
       },
 
+      reorder_visible: function() {
+          var visible = ''
+          var $itemList = $('.item-list');
+          var $visibleReports = $('.visible-reports');
+          for (var i = 0; i < fixmystreet.markers.features.length; i++) {
```
Comment (dracos) [inline web/js/map-OpenLayers.js]: I assume marker order here is the same as sidebar list order, might be worth a comment confirming that.

File: templates/web/base/report/_item.html [hunk 4]
```diff
@@ -27,7 +27,7 @@
 [% END %]
 
 <li class="item-list__item item-list--reports__item [% item_extra_class %]"
-    data-report-id="[% problem.id | html %]" data-lastupdate="[% problem.lastupdate %]">
+    data-report-id="[% problem.id | html %]" data-lastupdate="[% problem.lastupdate %]" id="report-[% problem.id | html %]">
```
Comment (dracos) [inline templates/web/base/report/_item.html]: Should whatever is using data-report-id be changed to use the ID instead?

File: web/js/map-OpenLayers.js [hunk 5]
```diff
@@ -115,6 +115,29 @@ var fixmystreet = fixmystreet || {};
...
+              var $listItem = $itemList.find('#report-'+ feature.data.id);
+              if (feature.onScreen()) {
+                  visible += $listItem[0].outerHTML; // This is faster than using .append()
+                  $listItem.addClass('hidden');
+              } else {
+                  $visibleReports.find('#report-'+ feature.data.id).remove();
+                  $listItem.removeClass('hidden');
+              }
+          }
+          if (visible == '') {
```
Comment (dracos) [inline web/js/map-OpenLayers.js]: Triple equals preferable here, if you're not going to have a boolean check.

File: web/js/map-OpenLayers.js (hunk 2 above)
Comment (pezholio) [inline web/js/map-OpenLayers.js]: Yeah, that was one of my concerns. I initially was selecting by `data-report-id`, but that was dog slow, which is why I added the ID. Wondering if I could append something to the ID when the element gets cloned and moved.

File: web/js/map-OpenLayers.js (hunk 1 above)
Comment (pezholio) [inline web/js/map-OpenLayers.js]: Ah, OK. I think you're right though. This behaviour is probably only appropriate on the `/reports` pages anyway

File: templates/web/base/report/_item.html (hunk 4 above)
Comment (pezholio) [inline templates/web/base/report/_item.html]: As I've now changed the IDs to be unique in the list, I think it's safer to have the `data` element here too

File: templates/web/base/reports/_problem-list.html [hunk 6]
```diff
@@ -3,7 +3,9 @@
 %]
 
 [% BLOCK column %]
-    <ul class="hidden item-list item-list--reports visible-reports"></ul>
+    [% IF reorder %]
+      <ul class="hidden item-list item-list--reports visible-reports"></ul>
```
Comment (dracos) [inline templates/web/base/reports/_problem-list.html]: More a musing than an issue, but is there a reason not to use an ID of, dunno, `js-visible-reports` here? To show it's for the JavaScript, and it's only being used once.

File: web/js/map-OpenLayers.js [hunk 7]
```diff
@@ -125,10 +125,13 @@ var fixmystreet = fixmystreet || {};
               var feature = fixmystreet.markers.features[i];
               var $listItem = $itemList.find('#report-'+ feature.data.id);
               if (feature.onScreen()) {
-                  visible += $listItem[0].outerHTML; // This is faster than using .append()
+                  var clone = $($listItem[0].outerHTML);
```
Comment (dracos) [inline web/js/map-OpenLayers.js]: Any reason not to use `$listItem.clone()`? :)

File: web/js/map-OpenLayers.js [hunk 8]
```diff
@@ -115,6 +115,34 @@ var fixmystreet = fixmystreet || {};
         fixmystreet.markers.redraw();
       },
 
+      reorder_visible: function() {
+          var visible = ''
+          var $itemList = $('.item-list');
+          var $visibleReports = $('.visible-reports');
+          if ($visibleReports.length == 0) return false;
```
Comment (dracos) [inline web/js/map-OpenLayers.js]: Please always have brackets with if statements (I don't think jslint will let this through anyway :) ).
At the start of this function, you could do `if fixmystreet.page != 'reports' { return; }` as that variable is always set, to save any DOM lookups at all.

File: web/js/map-OpenLayers.js [hunk 9]
```diff
@@ -115,6 +115,34 @@ var fixmystreet = fixmystreet || {};
...
+
+          for (var i = 0; i < fixmystreet.markers.features.length; i++) {
+              var feature = fixmystreet.markers.features[i];
+              var $listItem = $itemList.find('#report-'+ feature.data.id);
+              if (feature.onScreen()) {
+                  var clone = $($listItem[0].outerHTML);
```
Comment (dracos) [inline web/js/map-OpenLayers.js]: Any reason not to use `$listItem.clone()`? :)

File: templates/web/base/reports/_problem-list.html [hunk 10]
```diff
@@ -3,6 +3,9 @@
 %]
 
 [% BLOCK column %]
+    [% IF reorder %]
+      <ul class="hidden item-list item-list--reports visible-reports"></ul>
```
Comment (dracos) [inline templates/web/base/reports/_problem-list.html]: More a musing than an issue, but is there a reason not to use an ID of, dunno, `js-visible-reports here`? To show it's for the JavaScript, and it's only being used once.

File: web/js/map-OpenLayers.js (hunk 8 above)
Comment (pezholio) [inline web/js/map-OpenLayers.js]: Oh, that's a better solution 👍

File: templates/web/base/reports/_problem-list.html (hunk 10 above)
Comment (pezholio) [inline templates/web/base/reports/_problem-list.html]: Seems totally sensible and in line with the code style, so I'll do it 👍

File: templates/web/base/reports/_problem-list.html [hunk 11]
```diff
@@ -4,7 +4,7 @@
 
 [% BLOCK column %]
     [% IF reorder %]
-      <ul class="hidden item-list item-list--reports visible-reports"></ul>
+      <ul class="hidden item-list item-list--reports js-visible-reports"></ul>
```
Comment (dracos) [inline templates/web/base/reports/_problem-list.html]: This could be an ID, rather than a class.

**Comment by dracos** [comment]
This is no longer necessary, I'm afraid, due to #1689 `/reports` will now only show reports in view.

**Pull request closed without merging**